packages/camera/camera_avfoundation/ios/camera_avfoundation/Sources/camera_avfoundation/FLTCam.m

```diff
@@ -1302,9 +1303,47 @@ - (BOOL)setupWriterForPath:(NSString *)path {
   return YES;
 }
 
+// configure application wide audio session manually to prevent overwriting
+// flag MixWithOthers by capture session, only change category if it is considered
+// as upgrade which means it can only enable ability to play in silent mode or
+// ability to record audio but never disables it, that could affect other plugins
+// which depend on this global state, only change category or options if there is
+// change to prevent unnecessary route changes which can cause lags
+// https://github.com/flutter/flutter/issues/131553
+static void upgradeAudioSessionCategory(AVAudioSessionCategory category,
+                                        AVAudioSessionCategoryOptions options,
+                                        AVAudioSessionCategoryOptions clearOptions) {
+  if (!NSThread.isMainThread) {
+    dispatch_sync(dispatch_get_main_queue(), ^{
+      upgradeAudioSessionCategory(category, options, clearOptions);
```

Comment:
haven't looked into the details. Does this have to be on main queue? Can it be on capture session queue?

Reply:
Audio session category is global state, it is changed based on old state so it must be synchronized with other plugins changing it.

Reply:
There is needed `dispatch_sync`. Category must be `record` before adding inputs/outputs to capture session otherwise there would be no audio. It is called on the capture session queue so not on the main thread. Seems also `reportErrorMessage` is currently called only on capture session queue but there is that `FLTEnsureToRunOnMainQueue` anyway. If there was similar method which used `dispatch_sync` it would not make any difference it would just be `upgradeAudioSessionCategory->FLTEnsureToRunOnMainQueueSync->dispatch_sync` instead of `upgradeAudioSessionCategory->dispatch_sync->upgradeAudioSessionCategory` so the same depth in stack. Can it be even called recursion if that second call is in another thread on another stack? I think it would be ok to change it to just `dispatch_sync` as when this is called from the main thread it will crash with an unhandled exception instead of deadlock so that would be obvious for someone who did such a change. Btw are there better options on how to synchronize between independent plugins? Ideally it could be called on the original thread but with some mutex shared between these plugins (and something like `NSDistributedLock` seems like overkill).